High Score being Stored

hsavoie opened the pull request:
## Work Completed
- Created the `LocalData` class which provides static access to the shared preferences.
- Added methods to get/set the high score.
- Modified GameActivity & GameOverActivity to get/set the high score where needed.

hsavoie committed:
High Score being Stored

This is just a draft!!

I have the high score being stored, but I think I implemented it the opposite way. I think I should've done the work I did in GameOverActivity in GameActivity instead, then pass the high score value into GameOverActivity, rather than trying to pass the high score back into GameActivity from the value in GameOverActivity.

I tried to switch them up, but ran into some app crashes. Anyway, still working on it. (side note: Android Studio and my laptop do not get along so it's taking forever.)

fredy20501 commented on app/src/main/java/ca/unb/mobiledev/reflexrevolution/activities/GameOverActivity.java:
```diff
@@ -58,4 +68,19 @@ protected void onCreate(Bundle savedInstanceState) {
         menuButton.setOnClickListener(v -> finish());
     }
+
+    private void initSharedPreferences() {
+        prefs = getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
+    }
+
+    private int readHighScoreFromSharedPreferences() {
+        return prefs.getInt(HIGH_SCORE_KEY, 0);
+    }
+
+    private void writeHighScoreToSharedPreferences(int score) {
+        SharedPreferences.Editor editor = prefs.edit();
+        editor.putInt(HIGH_SCORE_KEY, score);
+        editor.apply();
+    }
```
It might be better to have these methods in their own separate static class. That way we can access the saved data from any activity very easily.

For example, we could have the static class like this:
```java
public class LocalData {
    private static final String PREFS_FILE_NAME = "AppPrefs";
    private static final String HIGH_SCORE_KEY = "HIGH_SCORE_KEY";
    private static SharedPreferences prefs;

    public static void initialize(Context context) {
        prefs = context.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
    }

    public static int getHighScore() {
        return prefs.getInt(HIGH_SCORE_KEY, 0);
    }

    public static void setHighScore(int score) {
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(HIGH_SCORE_KEY, score);
        editor.apply();
    }
}
```

So any activity can use it like so:
```java
LocalData.initialize(this);
int highScore = LocalData.getHighScore();
LocalData.setHighScore(100);
```

fredy20501 replied:
Also note that the high score is unique for each gamemode and difficulty level, so we would need to pass that information in when getting/setting the high score. We can use the names of the gamemode & difficulty to create the specific key for each combination. Something like this:
```java
public static int getHighScore(GameMode gameMode, Difficulty difficulty) {
    String key = HIGH_SCORE_KEY+"_"+gameMode.getName()+"_"+difficulty.getName();
    return prefs.getInt(key, 0);
}
```

carterjmoore replied:
I think Fred's suggestions seem like a great way to do it. Only thing I would add is maybe you would also have a private static method to get they key based on GameMode and Difficulty, which can then be re-used in the getHighScore() and setHighScore() methods. So like:

```java
private static int getHighScoreKey(GameMode gameMode, Difficulty difficulty) {
    return "HIGH_SCORE_KEY_"+gameMode.getName()+"_"+difficulty.getName();
}

public static int getHighScore(GameMode gameMode, Difficulty difficulty) {
    return prefs.getInt(getHighScoreKey(gameMode, difficulty), 0);
}

public static void setHighScore(int score, GameMode gameMode, Difficulty difficulty) {
    SharedPreferences.Editor editor = prefs.edit();
    editor.putInt(getHighScoreKey(gameMode, difficulty), score);
    editor.apply();
}
```

hsavoie committed:
Update High Score Storage

So I took your suggestions and I like it a lot more this way. However, when I try to set the text in GameActivity, the app still keeps crashing...

Not sure what I'm doing wrong here, I tried placing it in a few areas and confirmed the text box id, but I can't seem to figure out why it keeps crashing. Any suggestions? Ill continue to play around with it

hsavoie committed:
Finished High Score

- Changed initialize to initializeSP because there was already a method in GameActivity called initialize and this was the fastest way to fix it that I could think of.
- High score is displaying at the end and in the game activity, and is different for each game mode and difficulty

fredy20501 committed:
Code cleanup (minor changes)

fredy20501 reviewed (approved):
Nice work! 💯

carterjmoore reviewed (commented):
Looks good! 👍 💯 🥇

carterjmoore reviewed (approved):
My bad, approved this time 😅 🤣 😆
